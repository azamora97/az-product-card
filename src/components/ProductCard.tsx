import React, { createContext, CSSProperties, JSX } from 'react';
import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import { ProductContextProps, Product, onChangeArg, InitialValues, ProductCardHandlers } from '../interface/Interfaces';


export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product;
  children: ( args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArg) => void;
  value?: number;
  initialValues?: InitialValues
}


export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: Props) => {

  const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct( {onChange, product, value, initialValues} ); 
  
  return (
    <Provider value={{ 
        counter,
        product,
        maxCount,
        increaseBy
    }}>
        <div 
          className={ `${styles.productCard} ${className}` }
          style={ style }
        >
            { children({
                count: counter,
                isMaxCountReached,
                maxCount: initialValues?.maxCount,
                product,
            
                increaseBy,
                reset 
              }) 
            }
        </div>
    </Provider>
  )
}