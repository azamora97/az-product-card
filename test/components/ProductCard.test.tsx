import React from 'react';
import renderer from 'react-test-renderer';
import { product1 } from '../data/products';
import { ProductCard } from '../../src/components/index';

//const { act } = renderer;


describe('ProductCard', () => {

    test('debe de mostrar el componte de manera correcta', () => {


        const wrapper = renderer.create(
            <ProductCard product={product1}>
                {
                    () => (
                        <h1>Product Card</h1>
                    )
                }
            </ProductCard>
        )

        expect( wrapper.toJSON() ).toMatchSnapshot();
    })

    test('debe de incrementar el contador', () => {


        const wrapper = renderer.create(
            <ProductCard product={product1}>
                {
                    ({count, increaseBy}) => (
                        <>
                            <h1>Product Card</h1>
                            <span>{count}</span>
                            <button onClick={() => increaseBy(1)}>+1</button>
                        </>
                    )
                }
            </ProductCard>
        )

        let tree = wrapper.toJSON();
        expect( tree ).toMatchSnapshot();

        // no funciona ya que es obsoleto, se deberia usar como en el curso de react
       /* 
        act(() => {
            (tree as any).children[2].props.onClick();
        })
        tree = wrapper.toJSON();
        
        expect((tree as any).children[1].children).toBe(1);
        */
    })


})