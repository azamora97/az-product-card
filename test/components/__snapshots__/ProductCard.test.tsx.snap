// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard  1`] = `null`;

exports[`ProductCard debe de incrementar el contador 1`] = `null`;

exports[`ProductCard debe de mostrar el componte de manera correcta 1`] = `null`;
